src/Integration.Vsix/CFamily/CMake/CompilationDatabaseRequest.cs

```diff
@@ -0,0 +1,120 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.IO;
+using Newtonsoft.Json;
+using SonarLint.VisualStudio.CFamily.CMake;
+using SonarLint.VisualStudio.Core.CFamily;
+
+namespace SonarLint.VisualStudio.Integration.Vsix.CFamily.CMake
+{
+    /// <summary>
+    /// Encapsulates a low-level analysis request for the CFamily compilation database entry protocol
+    /// </summary>
+    internal class CompilationDatabaseRequest : IRequest
+    {
+        private readonly CompilationDatabaseEntry databaseEntry;
+
+        public CompilationDatabaseRequest(CompilationDatabaseEntry databaseEntry, RequestContext context)
+        {
+            this.databaseEntry = databaseEntry ?? throw new ArgumentNullException(nameof(databaseEntry));
+            Context = context ?? throw new ArgumentNullException(nameof(context));
+
+            // Must have a Command or Arguments but not both
+            var hasArgs = !string.IsNullOrEmpty(databaseEntry.Arguments);
+            var hasCmd = !string.IsNullOrEmpty(databaseEntry.Command);
+
+            if (!hasArgs && !hasCmd || (hasArgs && hasCmd))
+            {
+                throw new ArgumentException(CFamilyStrings.ERROR_InvalidCompilationEntry);
+            }
+        }
+
+        public RequestContext Context { get; }
+
+        public void WriteRequest(BinaryWriter writer)
+        {
+            Protocol.WriteUTF(writer, "SL-IN");
+
+            // Required inputs
+            WriteSetting(writer, "File", databaseEntry.File);
+            WriteSetting(writer, "Directory", databaseEntry.Directory);
+
+            if(databaseEntry.Arguments == null)
+            {
+                WriteSetting(writer, "Command", databaseEntry.Command);
+            }
+            else
+            {
+                WriteSetting(writer, "Arguments", databaseEntry.Arguments);
+            }
+
+            WriteQualityProfile(writer);
+
+            // Optional inputs
+            WriteSetting(writer, "PreambleFile", Context.PchFile);
+            WriteSetting(writer, "CreateReproducer", Context?.AnalyzerOptions?.CreateReproducer ?? false ? "true" : "false");
+            WriteSetting(writer, "BuildPreamble", Context?.AnalyzerOptions?.CreatePreCompiledHeaders ?? false ? "true" : "false");
+
+            // TODO - only supplied if analysing a header file
+//            WriteSetting(writer, "HeaderFilerLanguage", Context.CFamilyLanguage);
+            WriteRuleSettings(writer);
+
+            Protocol.WriteUTF(writer, "SL-END");
+        }
+
+        public void WriteRequestDiagnostics(TextWriter writer)
+        {
+            var data = JsonConvert.SerializeObject(databaseEntry, Formatting.Indented);
```

Comment:
I decided to indent the json to make it more readable. We could do the same for the other reproducer class.